pkg/skaffold/actions/runner.go

```diff
@@ -0,0 +1,152 @@
+	return Runner{execEnvByAction, orderedExecEnvs, acsByExecEnv}
+}
+
+func (r Runner) ExecAll(ctx context.Context, out io.Writer, allbuilds []graph.Artifact) error {
+	acs, err := r.prepareAllActions(ctx, out, allbuilds)
+	if err != nil {
+		return err
+	}
+
+	defer r.cleanup(ctx, out, acs, r.orderedExecEnvs)
+	return execWithFailingSafe(ctx, out, acs)
+}
+
+func (r Runner) Exec(ctx context.Context, out io.Writer, allbuilds []graph.Artifact, aName string) error {
+	execEnv, found := r.execEnvByAction[aName]
+	if !found {
+		return fmt.Errorf("custom action %v not found", aName)
+	}
+
+	acs, err := execEnv.PrepareActions(ctx, out, allbuilds, []string{aName})
+	if err != nil {
+		return err
+	}
+
+	// We expect only one action to be created.
+	if len(acs) != 1 {
+		return fmt.Errorf("failed to create %v action", aName)
+	}
+
+	a := acs[0]
+
+	defer r.cleanup(context.TODO(), out, []Task{a}, []ExecEnv{execEnv})
+	return a.Exec(ctx, out)
+}
+
+func (r Runner) prepareAllActions(ctx context.Context, out io.Writer, allbuilds []graph.Artifact) ([]Task, error) {
+	preparedAcs := []Task{}
+	for _, execEnv := range r.orderedExecEnvs {
+		acsNames := r.acsByExecEnv[execEnv]
+		acs, err := execEnv.PrepareActions(ctx, out, allbuilds, acsNames)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, a := range acs {
+			preparedAcs = append(preparedAcs, a)
+		}
+	}
+
+	return preparedAcs, nil
+}
+
+func (r Runner) cleanup(ctx context.Context, out io.Writer, ts []Task, execEnvs []ExecEnv) {
+	log.Entry(ctx).Debugf("Starting execution cleanup")
+	for _, t := range ts {
+		if t == nil {
+			continue
+		}
+		if err := t.Cleanup(ctx, out); err != nil {
+			// TODO(renzor): known issue related with Docker client deleting containers + prune will cause some
+			// warnings here. We need to fix https://github.com/GoogleContainerTools/skaffold/issues/8605
+			log.Entry(ctx).Warnf("%v cleanup error:%v", t.Name(), err)
+		}
+	}
+	for _, execEnv := range execEnvs {
+		if execEnv == nil {
+			continue
+		}
+		if err := execEnv.Cleanup(ctx, out); err != nil {
+			log.Entry(ctx).Warnf("Execution environment cleanup error:%v", err)
+		}
+	}
+}
+
+func getExecFunc(isFailFast bool) ExecStrategy {
+	if isFailFast {
+		return execWithFailingFast
+	}
+	return execWithFailingSafe
+}
+
+func execWithFailingFast(ctx context.Context, out io.Writer, ts []Task) error {
+	g, gCtx := errgroup.WithContext(ctx)
+
+	for _, t := range ts {
+		t := t
+		g.Go(func() error {
+			return execAndLog(gCtx, out, t)
+		})
+	}
+
+	return g.Wait()
+}
+
+func execWithFailingSafe(ctx context.Context, out io.Writer, ts []Task) error {
+	const maxWorkers = math.MaxInt64
+	g := semgroup.NewGroup(context.Background(), maxWorkers)
+
+	for _, t := range ts {
+		t := t
+		g.Go(func() error {
+			return execAndLog(ctx, out, t)
+		})
+	}
+
+	return g.Wait()
+}
+
+func execAndLog(ctx context.Context, out io.Writer, t Task) error {
+	log.Entry(context.TODO()).Debugf("Starting execution for %v", t.Name())
```

Comment:
can't `log.Entry` use `ctx` instead of `context.TODO`?

Reply:
I was thinking about a case when one action fails, canceling the context, but we should be fine using `ctx` here. Done, thanks!